native/src/modules/fetcher/FetcherModule.js

```diff
@@ -1,45 +1,55 @@
 // @flow
 
+import type { Channel } from 'redux-saga'
 import NativeFetcherModule, { NativeFetcherModuleEmitter } from './NativeFetcherModule'
+import { eventChannel, END } from 'redux-saga'
 import { isEmpty } from 'lodash'
 
 export type TargetFilePathsType = { [path: string]: string }
 
 export type FetchResultType = { [path: string]: {| lastUpdate: string, url: string, errorMessage: ?string |} }
 
-export type ProgressCallbackType = (progress: number) => void
-
 class FetcherModule {
   // TODO NATIVE-264: Correctly handle already fetching
   static currentlyFetching = false
 
-  async fetchAsync (
-    targetFilePaths: TargetFilePathsType,
-    progress: ProgressCallbackType
-  ): Promise<FetchResultType> {
+  fetchAsync = (targetFilePaths: TargetFilePathsType): [Promise<FetchResultType>, Channel] => {
     if (FetcherModule.currentlyFetching) {
       throw new Error('Already fetching!')
     }
-    if (isEmpty(targetFilePaths)) {
-      return {}
-    }
+
     FetcherModule.currentlyFetching = true
 
-    const subscriptions = []
-    subscriptions.push(NativeFetcherModuleEmitter.addListener('progress', progress))
+    const progressSubscription = NativeFetcherModuleEmitter.addListener('progress', emit)
 
+    let emit
+    const progressChannel = eventChannel<number>(emitter => {
+      emit = emitter
+      return () => {}
+    })
     try {
-      const result = await NativeFetcherModule.fetchAsync(targetFilePaths)
-
-      if (!result) {
-        // While testing the app I noticed that the FetchResultType of this function was empty. I am absolutely not sure
-        // why this happened. As this cause an inconsistent state it is better to throw an error in this case.
-        throw new Error('Fetch failed for some reason!')
+      var fetchPromise: Promise<FetchResultType>
+      if (isEmpty(targetFilePaths)) {
+        fetchPromise = Promise.resolve({})
+      } else {
+        fetchPromise = NativeFetcherModule.fetchAsync(targetFilePaths)
       }
 
-      return result
-    } finally {
-      subscriptions.forEach(sub => sub.remove())
+      fetchPromise.then(result => {
+        progressSubscription.remove()
+        emit(END)
+        FetcherModule.currentlyFetching = false
+
+        if (!result) {
+          // While testing the app I noticed that the FetchResultType of this function was empty. I am absolutely not sure
+          // why this happened. As this cause an inconsistent state it is better to throw an error in this case.
+          throw new Error('Fetch failed for some reason!')
+        }
+      })
+      return [fetchPromise, progressChannel]
+    } catch {
+      progressSubscription.remove()
+      emit(END)
       FetcherModule.currentlyFetching = false
     }
```

Comment:
I think I do not need the try catch block, but I'm afraid to remove it.

Reply:
But this was already the case before already that we don't have a catch block :thinking:

Reply:
I removed the catch block (previously "finally") so the "throw error" works again